signer/config.go

```diff
@@ -0,0 +1,281 @@
+package signer
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/legacy"
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	cryptocodec "github.com/cosmos/cosmos-sdk/crypto/codec"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/types/bech32"
+	"github.com/strangelove-ventures/horcrux/client"
+	"github.com/tendermint/tendermint/crypto"
+	"gopkg.in/yaml.v2"
+)
+
+type NodeConfig struct {
+	Address string
+}
+
+// Config maps to the on-disk JSON format
+type Config struct {
+	PrivValKeyFile *string         `json:"key-file,omitempty" yaml:"key-file,omitempty"`
+	ChainID        string          `json:"chain-id" yaml:"chain-id"`
+	CosignerConfig *CosignerConfig `json:"cosigner,omitempty" yaml:"cosigner,omitempty"`
+	ChainNodes     ChainNodes      `json:"chain-nodes,omitempty" yaml:"chain-nodes,omitempty"`
+	DebugAddr      string          `json:"debug-addr,omitempty" yaml:"debug-addr,omitempty"`
+}
+
+func (c *Config) Nodes() (out []string) {
+	for _, n := range c.ChainNodes {
+		out = append(out, n.PrivValAddr)
+	}
+	return out
+}
+
+func (c *Config) MustMarshalYaml() []byte {
+	out, err := yaml.Marshal(c)
+	if err != nil {
+		panic(err)
+	}
+	return out
+}
+
+func (c *Config) ValidateSingleSignerConfig() error {
+	if len(c.ChainNodes) == 0 {
+		return fmt.Errorf("need to have chain-nodes configured for priv-val connection")
+	}
+	if err := c.ChainNodes.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Config) ValidateCosignerConfig() error {
+	if err := c.ValidateSingleSignerConfig(); err != nil {
+		return err
+	}
+	if c.CosignerConfig == nil {
+		return fmt.Errorf("cosigner config can't be empty")
+	}
+	if c.CosignerConfig.Threshold <= c.CosignerConfig.Shares/2 {
+		return fmt.Errorf("threshold (%d) must be greater than number of shares (%d) / 2",
+			c.CosignerConfig.Threshold, c.CosignerConfig.Shares)
+	}
+	if c.CosignerConfig.Shares < c.CosignerConfig.Threshold {
+		return fmt.Errorf("number of shares (%d) must be greater or equal to threshold (%d)",
+			c.CosignerConfig.Shares, c.CosignerConfig.Threshold)
+	}
+
+	_, err := time.ParseDuration(c.CosignerConfig.Timeout)
+	if err != nil {
+		return fmt.Errorf("invalid --timeout: %w", err)
+	}
+	if _, err := url.Parse(c.CosignerConfig.P2PListen); err != nil {
+		return fmt.Errorf("failed to parse p2p listen address: %w", err)
+	}
+	if err := c.CosignerConfig.Peers.Validate(c.CosignerConfig.Shares); err != nil {
+		return err
+	}
+	return nil
+}
+
+type RuntimeConfig struct {
+	HomeDir    string
+	ConfigFile string
+	StateDir   string
+	PidFile    string
+	Config     Config
+}
+
+func (c RuntimeConfig) KeyFilePath(cosigner bool) string {
```

Comment:
I would split this into two different methods, instead of one method accepting a boolean parameter. When I saw `KeyFilePath(true)` earlier, I had no idea what true indicated, nor would I have been able to guess what false meant.

Some discussion on the topic: https://softwareengineering.stackexchange.com/q/147977